### ISSUE TITLE
Details on Opacity Loss

### ISSUE THREAD

**Issue by mvp18** [jaidevshriram/realmdreamer#4]
Hi,

Thanks for your great work! Could you provide some details on the opacity loss you mentioned, which is inspired by the Plenoxels paper? Since no equation is provided, I was hoping a short code snippet would help understand the formulation better. Thanks

**Comment by jaidevshriram** [comment]
Hi, Sure - it's similar to equation 5 in the Plenoxels paper in some sense - it's just a binary cross entropy loss on the opacity of individual points. Something like this:

```python            
# Opaqueness Loss
clamped_opacity = torch.clamp(self.gaussian_model.get_opacity, min=1e-5, max=1.0 - 1e-5)
loss_dict['loss_opaque'] = self.config.lambda_opaque * F.binary_cross_entropy(clamped_opacity, clamped_opacity)
```

**Comment by mvp18** [comment]
Thanks so much :)

**Comment by mvp18** [comment]
Hi,

Sorry, but I think there might be a bit of a typo in the code snippet you suggested. The loss would always default to zero as you have the same input and targets. Did you mean something like the following?

```
loss_dict['loss_opaque'] = -self.config.lambda_opaque * (torch.log(clamped_opacity) + torch.log(1 - clamped_opacity))
```

**Comment by jaidevshriram** [comment]
I believe my implementation is correct - the loss is minimized when the opacity is either 0 or 1, which can be seen when plugging in example values into the binary cross entropy equation. Can you clarify why it will be zero?

**Comment by mvp18** [comment]
Hi,

Sorry you are right it won't be zero. I forgot that F.binary_cross_entropy considers inputs with values between 0-1 while targets are usually 0/1 binary labels. But I think if you use F.binary_cross_entropy(clamped_opacity, clamped_opacity), you would be calculating
```
loss_dict['loss_opaque'] = -self.config.lambda_opaque * (clamped_opacity*torch.log(clamped_opacity) + (1-clamped_opacity)*torch.log(1 - clamped_opacity))
```

whereas going by Equation 5 in the Plenoxels paper, you mean to calculate what I wrote in my last comment. For example, for a tensor with all 0.7's, your equation would give self.config.lambda_opaque * 0.6109 whereas Equation 5 gives self.config.lambda_opaque * 1.5606 - more than double.

Hope that makes sense. Let me know if I wrote understanding something wrong.
